File: Visitor/c#/Visitor.cs
interface IVisitor
{
    void Visit(Zoo zoo);
    
    void Visit(Cinema cinema);
    
    void Visit(Circus circus);
}

interface IPlace
{
    void Accept(IVisitor v);
}

class Zoo : IPlace
{
    public void Accept(IVisitor v) => v.Visit(this);
}

class Cinema : IPlace
{
    public void Accept(IVisitor v) => v.Visit(this);
}

class Circus : IPlace
{
    public void Accept(IVisitor v) => v.Visit(this);
}

class HolidayMaker : IVisitor
{
    public string Value;

    public void Visit(Zoo zoo) => Value = "zoo";
    
    public void Visit(Cinema cinema) => Value = "cinema";
    
    public void Visit(Circus circus) => Value = "circus";
}

class Program
{
    static void Main()
    {
        List<IPlace> places = new List<IPlace>() {new Zoo(), new Cinema(), new Circus()};

        foreach (IPlace place in places)
        {
            HolidayMaker visitor = new HolidayMaker();
            place.Accept(visitor);
            Console.WriteLine(visitor.Value);
        }
    }
}

File: Visitor/java/Visitor.java
import java.util.ArrayList;
import java.util.List;

interface IVisitor{
    void visit(Zoo zoo);
    
    void visit(Cinema cinema);
    
    void visit(Circus circus);
}

interface IPlace{
    void accept(IVisitor v);
}

class Zoo implements IPlace{
    @Override
    public void accept(IVisitor v){
        v.visit(this);
    }
}

class Cinema implements IPlace{
    @Override
    public void accept(IVisitor v){
        v.visit(this);
    }
}

class Circus implements IPlace{
    @Override
    public void accept(IVisitor v){
        v.visit(this);
    }
}

class HolidayMaker implements IVisitor {
    public String value;

    @Override
    public void visit(Zoo zoo) {
        value = "zoo";
    }

    @Override
    public void visit(Cinema cinema) {
        value = "cinema";
    }

    @Override
    public void visit(Circus circus) {
        value = "circus";
    }
}

public class Main{
    public static void main(String[] args){

        List<IPlace> places = new ArrayList<>();

        places.add(new Zoo());
        places.add(new Cinema());
        places.add(new Circus());

        for (IPlace place : places){
            HolidayMaker visitor = new HolidayMaker();
            place.accept(visitor);
            System.out.println(visitor.value);
        }
    }
}